diff -urN a/board/nedap/nedap9263/nedap9263.c b/board/nedap/nedap9263/nedap9263.c
--- a/board/nedap/nedap9263/nedap9263.c	2010-12-02 11:44:50.933585723 +0100
+++ b/board/nedap/nedap9263/nedap9263.c	2010-12-02 11:49:10.172897341 +0100
@@ -63,15 +63,15 @@
 	writel(csa, &matrix->csa[0]);
 
 	/* Configure SMC CS3 for NAND/SmartMedia */
-	writel(AT91_SMC_SETUP_NWE(1) | AT91_SMC_SETUP_NCS_WR(1) |
-		AT91_SMC_SETUP_NRD(1) | AT91_SMC_SETUP_NCS_RD(1),
+	writel(AT91_SMC_SETUP_NWE(2) | AT91_SMC_SETUP_NCS_WR(0) |
+		AT91_SMC_SETUP_NRD(2) | AT91_SMC_SETUP_NCS_RD(0),
 		&smc->cs[3].setup);
 
-	writel(AT91_SMC_PULSE_NWE(3) | AT91_SMC_PULSE_NCS_WR(3) |
-		AT91_SMC_PULSE_NRD(3) | AT91_SMC_PULSE_NCS_RD(3),
+	writel(AT91_SMC_PULSE_NWE(4) | AT91_SMC_PULSE_NCS_WR(4) |
+		AT91_SMC_PULSE_NRD(4) | AT91_SMC_PULSE_NCS_RD(4),
 		&smc->cs[3].pulse);
 
-	writel(AT91_SMC_CYCLE_NWE(5) | AT91_SMC_CYCLE_NRD(5),
+	writel(AT91_SMC_CYCLE_NWE(7) | AT91_SMC_CYCLE_NRD(7),
 		&smc->cs[3].cycle);
 	writel(AT91_SMC_MODE_RM_NRD | AT91_SMC_MODE_WM_NWE |
 		AT91_SMC_MODE_EXNW_DISABLE |
@@ -80,7 +80,7 @@
 #else /* CONFIG_SYS_NAND_DBW_8 */
 		       AT91_SMC_MODE_DBW_8 |
 #endif
-		       AT91_SMC_MODE_TDF_CYCLE(2),
+		       AT91_SMC_MODE_TDF_CYCLE(3),
 		&smc->cs[3].mode);
 
 	writel(1 << AT91SAM9263_ID_PIOA | 1 << AT91SAM9263_ID_PIOCDE,
@@ -194,11 +194,11 @@
 		AT91_SMC_SETUP_NRD(0) | AT91_SMC_SETUP_NCS_RD(0),
 		&smc->cs[0].setup);
 
-	writel(AT91_SMC_PULSE_NWE(5) | AT91_SMC_PULSE_NCS_WR(6) |
-		AT91_SMC_PULSE_NRD(3) | AT91_SMC_PULSE_NCS_RD(6),
+	writel(AT91_SMC_PULSE_NWE(7) | AT91_SMC_PULSE_NCS_WR(8) |
+		AT91_SMC_PULSE_NRD(3) | AT91_SMC_PULSE_NCS_RD(8),
 		&smc->cs[0].pulse);
 
-	writel(AT91_SMC_CYCLE_NWE(6) | AT91_SMC_CYCLE_NRD(6),
+	writel(AT91_SMC_CYCLE_NWE(8) | AT91_SMC_CYCLE_NRD(8),
 		&smc->cs[0].cycle);
 		       
 	/* test PSRAM */
